velox/docs/develop/debugging/print-expr-with-stats.rst

```diff
@@ -0,0 +1,148 @@
+column that uses the same base dictionary, the evaluation remembers the results
+for the dictionary rows and evaluates only on the new rows.
+
+For example, we evaluated log2(c0) * 1.5 expressions on 3 batches of data. Each
+batch had 1024 rows dictionary encoded using the same dictionary. The
+dictionary also has 1024 rows. First batch referenced first ⅕ of the dictionary
+rows repeating each 5 times. Second batch referenced the first  ⅓ of the
+dictionary rows repeating each 3 times. Third batch referenced the first 1/10
+of the dictionary rows repeating each 10 times.
+
+After evaluating the first batch printExprWithStats showed that expression was
+evaluated on 205 rows ( ⅕ of 1024).
+
+.. code-block::
+
+    multiply [cpu time: 29.76us, rows: 205] -> DOUBLE [#1]
+       log2 [cpu time: 41.49us, rows: 205] -> DOUBLE [#2]
+          cast(c0 as DOUBLE) [cpu time: 93.22us, rows: 205] -> DOUBLE [#3]
+             c0 [cpu time: 0ns, rows: 0] -> INTEGER [#4]
+       1.5:DOUBLE [cpu time: 0ns, rows: 0] -> DOUBLE [#5]
+
+After evaluating the second batch printExprWithStats showed that expression was
+evaluated on 137 additional rows (137 = ⅓ - ⅕ of 1024). This is expected as we
+have already computed the results for the first 205 rows and only need to
+compute the results for the remaining 137 rows.
+
+.. code-block::
+
+    multiply [cpu time: 44.13us, rows: 342] -> DOUBLE [#1]
+       log2 [cpu time: 69.13us, rows: 342] -> DOUBLE [#2]
+          cast(c0 as DOUBLE) [cpu time: 124.33us, rows: 342] -> DOUBLE [#3]
+             c0 [cpu time: 0ns, rows: 0] -> INTEGER [#4]
+       1.5:DOUBLE [cpu time: 0ns, rows: 0] -> DOUBLE [#5]
+
+After evaluating the third batch printExprWithStats shows that expression was
+not evaluated on any additional rows. This is expected as we have already
+computed the results on all the necessary rows.
+
+.. code-block::
+
+    multiply [cpu time: 44.13us, rows: 342] -> DOUBLE [#1]
+       log2 [cpu time: 69.13us, rows: 342] -> DOUBLE [#2]
+          cast(c0 as DOUBLE) [cpu time: 124.33us, rows: 342] -> DOUBLE [#3]
+             c0 [cpu time: 0ns, rows: 0] -> INTEGER [#4]
+       1.5:DOUBLE [cpu time: 0ns, rows: 0] -> DOUBLE [#5]
+
+ExprSetListener
+---------------
+
+One can register a listener that is invoked on destruction of the ExprSet and
+receives the aggregated runtime statistics for the whole expression tree. The
+statistics of the individual expressions as combined based on the expression
+name, i.e. function name or built-in expression name such as AND, OR, IF,
+SWITCH, etc. The listener can be used to log the runtime stats for offline
+analysis.
+
+.. code-block:: c++
+
+    /// Listener invoked on ExprSet destruction.
+    class ExprSetListener {
+     public:
+      virtual ~ExprSetListener() = default;
+
+      /// Called on ExprSet destruction. Provides runtime statistics about
+      /// expression evaluation.
+      /// @param uuid Universally unique identifier of the set of expressions.
```

Comment:
How is the uuid generated ? For a given example with multiple expression sets how can we get the association between the expression and the uuid ?

Reply:
UUID is auto-generated for each instance of ExprSet, which represents one or more expressions.

```
  ExprSet(
      std::vector<std::shared_ptr<const core::ITypedExpr>>&& source,
      core::ExecCtx* execCtx,
      bool enableConstantFolding = true);
...

ExprSet::~ExprSet() {
  listeners().withRLock([&](auto& listeners) {
    if (!listeners.empty()) {
      std::unordered_map<std::string, exec::ExprStats> stats;
      for (const auto& expr : exprs()) {
        addStats(*expr, stats);
      }

      auto uuid = makeUuid();
      for (auto& listener : listeners) {
        listener->onCompletion(uuid, {stats});
      }
    }
  });
}
```